autofit/graphical/laplace/newton.py

```diff
@@ -0,0 +1,262 @@
+    Bzk: AbstractVariableOperator,
+):
+    dk = VariableData.sub(xk, xk1)
+    zk = VariableData.sub(fk.deterministic_values, fk1.deterministic_values)
+    zkTzk2 = zk.dot(zk) ** 2
+    alpha = (zk.dot(Bzk.dot(zk)) - dk.dot(Bxk.dot(dk))) / zkTzk2
+    return Bzk.update((zk, alpha * zk))
+
+
+## Newton step
+
+
+class NextStep(NamedTuple):
+    fval: FactorValue
+    xval: VariableData
+    gval: VariableData
+    f_count: int
+    g_count: int
+
+
+class NextQuasiStep(NamedTuple):
+    fval: FactorValue
+    xval: VariableData
+    gval: VariableData
+    B: AbstractVariableOperator
+    det_B: Optional[AbstractVariableOperator]
+    f_count: int
+    g_count: int
+
+
+def take_step(
+    factor: FactorInterface,
+    factor_jacobian: FactorJacobianInterface,
+    xk: VariableData,
+    gk: Optional[VariableData] = None,
+    bk: Optional[AbstractVariableOperator] = None,
+    fk: Optional[FactorValue] = None,
+    fk_m1: Optional[FactorValue] = None,
+    *,
+    args: tuple = (),
+    search_direction=newton_direction,
+    calc_line_search: LineSearchFunc = line_search,
+    search_direction_kws: Optional[Dict[str, Any]] = None,
+    line_search_kws: Optional[Dict[str, Any]] = None,
+    f_count=0,
+    g_count=0,
+):
+    ls_kws = line_search_kws or {}
+    if gk is None:
+        _, gk = factor_jacobian(xk, *args)
+        g_count += 1
+
+    pk = search_direction(xk, gk, bk, **(search_direction_kws or {}))
+    ls = calc_line_search(
+        factor,
+        factor_jacobian,
+        xk,
+        pk,
+        gk,
+        fk,
+        fk_m1,
+        args=args,
+        f_count=f_count,
+        g_count=g_count,
+        **ls_kws,
+    )
+    x_next = xk + VariableData.mul(pk, ls.alpha)
+    return NextStep(ls.fval, x_next, ls.gval, ls.f_count, ls.g_count)
+
+
+def take_quasi_newton_step(
+    factor: FactorInterface,
+    factor_jacobian: FactorJacobianInterface,
+    xk: VariableData,
+    gk: Optional[VariableData] = None,
+    Bk: Optional[AbstractVariableOperator] = None,
+    fk: Optional[FactorValue] = None,
+    fk_m1: Optional[FactorValue] = None,
+    det_Bk: Optional[AbstractVariableOperator] = None,
+    *,
+    args: tuple = (),
+    search_direction=newton_direction,
+    calc_line_search: LineSearchFunc = line_search,
+    quasi_newton_update=bfgs_update,
+    search_direction_kws: Optional[Dict[str, Any]] = None,
+    line_search_kws: Optional[Dict[str, Any]] = None,
+    quasi_newton_kws: Optional[Dict[str, Any]] = None,
+    f_count=0,
+    g_count=0,
+) -> NextQuasiStep:
+    next_step = take_step(
+        factor,
+        factor_jacobian,
+        xk,
+        gk,
+        Bk,
+        fk,
+        fk_m1,
+        args=args,
+        search_direction=search_direction,
+        calc_line_search=calc_line_search,
+        line_search_kws=line_search_kws,
+        search_direction_kws=search_direction_kws,
+        f_count=f_count,
+        g_count=g_count,
+    )
+    Bk_next = quasi_newton_update(
```

Comment:
The uppercase variables represent matrices, it's more confusing from a mathematical notation perspective to have matrices represented by lower case variables unfortunately.

Reply:
Yeah I see that. I guess we should do whatever is best in terms of interpretting the code. From a Python perspective that would be strictly adhering to Python conventions but I can see how useful it could be to use cases in a more mathematical function.